### PR TITLE
feat(lyrics-plus): save loaded lyrics & support for enhanced LRC

### PR DESCRIPTION
**Pull request by kyrie25**
In request of #2158.
Track link for cross-referrence: [505 - Arctic Monkeys](https://open.spotify.com/track/0BxE4FqsDD1Ot4YuBXwAPp?si=b513cf4d07f6469d)

Spotify_eydpCXIjBF

### CODE REVIEW

**Comment by kyrie25** [comment]
Some more stuff

image
image

**Comment by kyrie25** [comment]
Should be stable now.

File: CustomApps/lyrics-plus/Settings.js [hunk 1]
```diff
@@ -37,6 +37,35 @@ const SwapButton = ({ icon, disabled, onClick }) => {
...
+
+	return react.createElement(
+		"button",
+		{
+			className: "btn",
+			onClick: () => {
+				localStorage.removeItem("lyrics-plus:local-lyrics");
+				setCount(0);
+			},
+			disabled: count === 0
```
Comment (rxri) [inline CustomApps/lyrics-plus/Settings.js:63]:
```suggestion
			disabled: !count
```

File: CustomApps/lyrics-plus/Settings.js [hunk 2]
```diff
@@ -37,6 +37,35 @@ const SwapButton = ({ icon, disabled, onClick }) => {
...
+		if (!localLyrics || typeof localLyrics !== "object") {
+			throw "";
+		}
+		lyrics = localLyrics;
+	} catch {
+		lyrics = {};
+	}
+
+	const [count, setCount] = useState(Object.keys(lyrics).length);
+	const text = count > 0 ? "Clear cached lyrics" : "No cached lyrics";
```
Comment (rxri) [inline CustomApps/lyrics-plus/Settings.js:53]:
```suggestion
	const text = !count ? "Clear cached lyrics" : "No cached lyrics";
```

File: CustomApps/lyrics-plus/Utils.js [hunk 3]
```diff
@@ -82,39 +82,117 @@ class Utils {
...
+	static formatTime(timestamp) {
+		if (!isNaN(timestamp)) {
+			let minutes = Math.trunc(timestamp / 60000),
+				seconds = ((timestamp - minutes * 60000) / 1000).toFixed(2);
+
+			if (minutes < 10) minutes = "0" + minutes;
+			if (seconds < 10) seconds = "0" + seconds;
+
+			return `${minutes}:${seconds}`;
+		} else return timestamp.toString();
```
Comment (rxri) [inline CustomApps/lyrics-plus/Utils.js:94]:
```suggestion
		if (isNaN(timestamp)) return timestamp.toString();
		let minutes = Math.trunc(timestamp / 60000),
			seconds = ((timestamp - minutes * 60000) / 1000).toFixed(2);

		if (minutes < 10) minutes = "0" + minutes;
		if (seconds < 10) seconds = "0" + seconds;

		return `${minutes}:${seconds}`;
```

File: CustomApps/lyrics-plus/Utils.js [hunk 4]
```diff
@@ -82,39 +82,117 @@ class Utils {
...
+		function formatTime(timestamp) {
+			if (!isNaN(timestamp)) {
+				let minutes = Math.trunc(timestamp / 60000),
+					seconds = ((timestamp - minutes * 60000) / 1000).toFixed(2);
+
+				if (minutes < 10) minutes = "0" + minutes;
+				if (seconds < 10) seconds = "0" + seconds;
+
+				return `${minutes}:${seconds}`;
+			} else return timestamp.toString();
```
Comment (rxri) [inline CustomApps/lyrics-plus/Utils.js:107]:
```suggestion
		if (isNaN(timestamp)) return timestamp.toString();
		let minutes = Math.trunc(timestamp / 60000),
			seconds = ((timestamp - minutes * 60000) / 1000).toFixed(2);
		if (minutes < 10) minutes = "0" + minutes;
		if (seconds < 10) seconds = "0" + seconds;
		return `${minutes}:${seconds}`;
```

File: CustomApps/lyrics-plus/Utils.js [hunk 5]
```diff
@@ -82,39 +82,117 @@ class Utils {
...
+		const karaoke = isKaraoke ? [] : null;
+
+		function timestampToMs(timestamp) {
+			const [minutes, seconds] = timestamp.replace(/\[\]\<\>/, "").split(":");
+			return Number(minutes) * 60 * 1000 + Number(seconds) * 1000;
+		}
+
+		function parseKaraokeLine(line, startTime) {
+			let wordTime = timestampToMs(startTime);
+			let karaokeLine = [];
```
Comment (rxri) [inline CustomApps/lyrics-plus/Utils.js:165]:
```suggestion
			const karaokeLine = [];
```

**Review by rxri — CHANGES_REQUESTED** [review]

**Review by rxri — APPROVED** [review]

**Review by theRealPadster — APPROVED** [review]
LGTM

**Pull request merged**